x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -1413,6 +1414,28 @@ export function isReferenced(layer: IndexPatternLayer, columnId: string): boolea
   return allReferences.includes(columnId);
 }
 
+const computeReferenceLookup = memoizeOne((layer: IndexPatternLayer): Record<string, string> => {
+  // speed up things for deep chains as in formula
+  const refLookup: Record<string, string> = {};
+  for (const [parentId, col] of Object.entries(layer.columns)) {
+    if ('references' in col) {
+      for (const colId of col.references) {
+        refLookup[colId] = parentId;
+      }
+    }
+  }
+  return refLookup;
+});
+
+export function getTopColumnFromReference(layer: IndexPatternLayer, columnId: string): string {
```

Comment:
maybe `getReferencedColumnsRoot`? `getReferenceHead`? Not sure what the best name is here. Maybe a little explaining comment above the function would be helpful.